collect_app/src/main/java/org/odk/collect/android/architecture/MVVMActivity.java

```diff
@@ -0,0 +1,83 @@
+package org.odk.collect.android.architecture;
+
+
+import android.arch.lifecycle.ViewModelProvider;
+import android.arch.lifecycle.ViewModelProviders;
+import android.os.Bundle;
+import android.support.annotation.LayoutRes;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import org.odk.collect.android.activities.InjectableActivity;
+
+import javax.inject.Inject;
+
+import butterknife.ButterKnife;
+import butterknife.Unbinder;
+
+/**
+ * A new Activity base class that uses Dagger to bootstrap VM creation.
+ *
+ * Create a new {@link MVVMViewModel} subclass, override getViewModelClass() to
+ * return the subclass you've created, and you'll have access to a persistent
+ * VM in your onCreate thanks to Dagger.
+ *
+ * Also handles subclass and Fragment injection, just add {@link Inject} fields
+ * to your Activity and have any Fragments you want injected implement the
+ * {@link Injectable} interface.
+ *
+ * @param <V> The MVVMViewModel subclass this Activity should load.
+ *
+ */
+public abstract class MVVMActivity<V extends MVVMViewModel>
+        extends InjectableActivity {
+
+    @Inject
+    protected ViewModelProvider.Factory viewModelFactory;
+
+    @Nullable
+    private V viewModel;
+
+    @Nullable
+    private Unbinder unbinder;
+
+    @Override
+    protected void onCreate(@Nullable Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(getLayoutId());
+
+        viewModel = ViewModelProviders.of(this, viewModelFactory).get(getViewModelClass());
+        viewModel.create();
+
+        unbinder = ButterKnife.bind(this);
+    }
+
+    @Override
+    protected void onDestroy() {
+        super.onDestroy();
+
+        if (unbinder != null) {
+            unbinder.unbind();
+            unbinder = null;
+        }
+
+        viewModel = null;
+        viewModelFactory = null;
+    }
+
+    @NonNull
+    public V getViewModel() {
+        if (viewModel == null) {
```

Comment:
I'm curious on how often this will actually occur. As far as I can tell, the only time that a child Activity (Say, GeoActivity) would have a null viewModel is if it called `getViewModel()` _before_ it called `super.onCreate()`

```
@Override
protected void onCreate(@Nullable Bundle savedInstanceState) {
    getViewModel().subscribe()
    super.onCreate(savedInstanceState);
}
```

I guess it is possible for it to be null when the activity changes orientation or comes back into the foreground? But then I believe `onCreate()` is called again. Nothing to get too hung up on, but I'd like to know the thought behind this decision.

Reply:
It should *never* occur, but I think its important to mark methods `@NonNull`/`@Nullable` and since the actual field technically *is* nullable, this was the best way to keep it `@NonNull`.

Reply:
👍 I missed that `viewModel` was declared nullable and that the method was not.